ext/cl_khr_command_buffer.asciidoc

```diff
@@ -452,8 +453,12 @@ Add to *Table 5*, _Device Queries_, of section 4.2:
   `CL_COMMAND_BUFFER_CAPABILITY_SIMULTANEOUS_USE_KHR` Device supports the
   command-buffers having a <<pending_count, Pending Count>> that exceeds 1.
 
-  `CL_COMMAND_BUFFER_CAPABILITY_OUT_OF_ORDER_KHR` Device supports the ability
-  to record command-buffers to out-of-order command-queues.
+| `CL_DEVICE_COMMAND_BUFFER_SUPPORTED_QUEUE_PROPERTIES_KHR`
+| `cl_command_queue_properties`
+| Bitmask of the supported properties with which a command-queue may be created
+  to allow a command-buffer to be executed on it. It is invalid for a
+  command-queue to be created with a property not reported and still be
+  compatible with command-buffer execution.
```

Comment:
This mechanism will work fine for command-queue properties that are expressed via the `cl_command_queue_properties` bitfield but it won't work as well for command-queue properties that are expressed via `cl_queue_properties` property-value pairs and passed to `clCreateCommandQueueWithProperties`.  Are we OK with this?

Regardless, are there some command-queue properties that must be supported by all implementations?  For example, would we expect all implementations to support `CL_QUEUE_PROFILING_ENABLE` because this is required by [`CL_DEVICE_QUEUE_ON_HOST_PROPERTIES`](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#CL_DEVICE_QUEUE_ON_HOST_PROPERTIES)?

Reply:
> This mechanism will work fine for command-queue properties that are expressed via the cl_command_queue_properties bitfield but it won't work as well for command-queue properties that are expressed via cl_queue_properties property-value pairs and passed to clCreateCommandQueueWithProperties. Are we OK with this?

That's a good point, we should check that the WG is OK with this. We already return `cl_command_queue_properties` for `CL_DEVICE_COMMAND_BUFFER_REQUIRED_QUEUE_PROPERTIES_KHR`, so the spec has this limitation in another place. We could return a `cl_queue_properties`  from these queries rather than `cl_command_queue_properties`, but it would make the extension incompatible with 1.2 implementations, which would be a shame.

I feel like this is limitation is okay, as we can still add a bespoke `cl_device_command_buffer_capabilities_khr` capability for a `cl_queue_properties ` pair that can't be represented by a property if the situation comes up. So I don't think this changes makes things any worse for this scenario.


> Regardless, are there some command-queue properties that must be supported by all implementations? For example, would we expect all implementations to support CL_QUEUE_PROFILING_ENABLE because this is required by CL_DEVICE_QUEUE_ON_HOST_PROPERTIES?

This update would indeed remove any baseline properties that a user could rely on across all implementations. We are actually finding it non-trivial to implement command-buffer profiling in ComputeAorta properly without the profiling information from our counters being overwritten by subsequent enqueues of the same command-buffer. So would not report support for `CL_QUEUE_PROFILING_ENABLE` from `CL_DEVICE_COMMAND_BUFFER_SUPPORTED_QUEUE_PROPERTIES_KHR` until we ironed out those issues if the spec made that possible.

Reply:
Selfishly, having mandatory support for event profiling is really nice because it enables standard and portable profilers like the OpenCL intercept layer.  I suppose it's worth observing though that profiling command buffers is one of the few features that is not currently supported by the command buffer emulation layer, so I can certainly appreciate some of the complexity.

Note that the problem for profilers may be even worse than it originally appears depending what it means if a command buffer does not support profiling: At least for the OpenCL intercept layer, when device profiling is enabled **all** command queues are created with profiling enabled.  If command buffers do not support profiling would this mean that these command queues wouldn't be able to execute command buffers at all?  If so, that would be a pretty painful restriction and it'd be tricky to work around.

FWIW, here is how I was thinking of doing event profiling in the command buffer emulation layer if this ends up being required.  This is the general case that works for out-of-order queues, but it may be simplified slightly for an in-order queue, or made a little more complex to use a marker instead of a barrier:

```c
// sequence to emulate a call to:
// clEnqueueCommandBuffer(1, &queue, cmdbuf, num_events, wait_list, &event)

cl_event start;
clEnqueueBarrierWithWaitList(queue, num_events, wait_list, &start);

// execute all of the commands in cmdbuf here:
// clEnqueueSomeCommand(queue, ...);

clEnqueueBarrierWithWaitList(queue, 0, NULL, &event);

// for the purposes of event profiling:
// * use the "start" event for QUEUED, SUBMIT, and START
// * use the "end" event for END and COMPLETE
```

This doesn't seem too bad?

Reply:
hmm I see the issue with the OpenCL intercept layer, I think we should specify that `CL_QUEUE_PROFILING_ENABLE` must be reported then, which preserves the current command-buffer spec behaviour. I can push a commit to update this PR to that.

Your proposed emulation layer implementation makes sense. We do something similar with the command-buffer objects
in our underlying API, it's just propagating the logic for working with the equivalent of start/end events that gets tricky. It's definitely achievable though, so if mandating profiling support improves the user experience then we should go with that.